# 예외처리

### 개요

> 프로그램에서 오류가 발생했을 때 로직을 다르게 처리하는 조치

**예외처리의 목적**

1. 예외의 발생으로 인한 실행 중인 프로그램의 비정상 종료를 막기 위해
2. 개발자에게 알려서 코드를 보완할 수 있도록 하게 위해

### Java의 예외처리

#### try-catch(-finally) 형식

**템플릿**

```java
try {
    // 예외가 발생할 가능성이 있는 코드를 구현합니다.
} catch (FileNotFoundException e) {
    // FileNotFoundException이 발생했을 경우,이를 처리하기 위한 코드를 구현합니다.
} catch (IOException e) {
    // FileNotFoundException이 아닌 IOException이 발생했을 경우,이를 처리하기 위한 코드를 구현합니다.
} finally {
    // 예외의 발생여부에 관계없이 항상 수행되어야하는 코드를 구현합니다.
}
```

- `finally` 구문은 필수는 아니다.

#### try-with-resource 형식

- 입출력 시에 쓰이는 예외처리 구문

```java
import java.io.FileOutputStream;
import java.io.IOException;

public class Main {
    public static void main(String[] args) {

        try (FileOutputStream out = new FileOutputStream("test.txt")) {
            // test.txt file 에 Hello Sparta 를 출력
            out.write("Hello Sparta".getBytes());
            out.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

#### 메소드에서의 예외 선언

- catch문을 이용해서 예외처리를 하지 않은 경우, 메소드에 `throws`로 예외가 발생할 수 있다는 것을 알려 준다.

```java
void method() throws IndexOutOfBoundsException, IllegalArgumentException {
    //메소드의 내용
}
```

